### PR TITLE
[bugfix] improves behavior of session/receiver/sender creation methods during cancellation

### PR DESCRIPTION
**Pull request by chradek**
## Description

When createSession, createReceiver, or createSender are cancelled via an abortSignal, the underlying rhea analogue is closed, regardless of if it has actually been opened.

In the `@azure/event-hubs` library, I ran into an edge case caused by this behavior. In `@azure/event-hubs`, I had a scenario where I'd create multiple receivers expecting an error from the service, and then triggered the abortSignal as soon as the 1st receiver was created. Any receivers that weren't opened prior to the abortSignal being triggered would cause the service to respond with an error stating they could not find the receiver's session handle in the connection.

This error from the service makes sense because from the service's point of view, it never saw a session for the receiver created.

This PR adds a check before closing the underlying rhea resource that ensures it is actually open 1st.

## Edit:

I updated the logic around when to close the rhea session/receiver/sender when cancelling their creation.

Through some more testing I found that each resource's `open` event could still be emitted after aborting even if `resource.is_open()` returned false. As pointed out in the comments, is_open() returns false if the local is open but the remote is not. So what's happening is that the begin/attach frame is send to the peer, we abort, then the peer responds saying it began/attached.

So now I check if the resource is already closed. If it isn't, I attach a listener for the resource's `open` event and call close on that resource if the event is triggered. This ensures that the resource is removed from its parent's state.

I've updated the tests to account for this and ensure that the resources are properly removed from their parent. I also found that in the "pre-cancelled" cases, we were creating the resources and the aborting, rather than just aborting right away, so I've updated this as well.

### CODE REVIEW

File: lib/connection.ts [hunk 1]
```diff
@@ -577,7 +577,9 @@ export class Connection extends Entity {
 
       onAbort = () => {
         removeListeners();
-        rheaSession.close();
+        if (rheaSession.is_open()) {
+          rheaSession.close();
```
Comment (ramya-rao-a) [inline lib/connection.ts:581]: Why wouldnt rhea consider this to be a no-op if the session is not open in the first place?
Is it because `is_open` checks for both local and remote open while `close()` ends up in https://github.com/amqp/rhea/blob/master/lib/endpoint.js#L48 where we check only for local open?

Comment (chradek) [reply lib/connection.ts:581]: Right. Because close doesn't check if the remote is open, we end up trying to close something that hasn't actually been opened on the remote side yet.

Comment (ramya-rao-a) [reply lib/connection.ts:581]: How do we ensure "local" stuff get cleaned up in such cases?

Comment (ramya-rao-a) [reply lib/connection.ts:581]: @grs What is your recommendation here? Should we always check `is_open()` before attempting a `close()` or should `close()` be smart enough to not communicate to the server to close something that is not open?

Comment (chradek) [reply lib/connection.ts:581]: > How do we ensure "local" stuff get cleaned up in such cases?

I could call `remove()` on the rhea receiver/sender, which should remove it from the session. There is a note that this shouldn't be done on active links, but I don't think the aborted links should be considered active.

I can do some more testing to see what happens if I abort these without closing the connection (in our libraries we do close the connection so nothing is left open).

Comment (grs) [reply lib/connection.ts:581]: The close() does involve a check on whether the endpoint is locally open or not, so I don't believe a check is necessary to prevent a close for a session that has already been opened and closed. Closing locally is fine even if remote has not yet signalled open. What would not be ok is sending a close when no open has been sent (i.e. calling close() if open()/begin() was never called). The only way I see that could happen is if the session object was never opened locally.

The session returned by rhea from Connection.create_session() is not locally opened. However the links returned from open_sender/open_receiver are opened.

Comment (richardpark-msft) [reply lib/connection.ts:581]: I think the local workaround is fine for now, although it'd be nice to put it into a function and put a comment on it to indicate why it's needed (I can easily see someone thinking - oh, we don't need to bother and never be able to trigger the edge case!).

Also, can we add a test for this?

Comment (chradek) [reply lib/connection.ts:581]: So an alternative I'm thinking about for receiver and sender (might work for session as well) is instead of calling `rheaReceiver.close()` right away in our `onAbort` handler, I can attach a "receiver_open" event listener on the `rheaReceiver` object. If it is invoked, this would mean the receiver was opened after we initiated our abort, and at this point we just close the receiver.

We could also have the is_open check I've added and close if it's true, since at that point we know the remote is open as well.

File: lib/connection.ts [hunk 2]
```diff
@@ -533,14 +533,27 @@ export class Connection extends Entity {
...
     return new Promise((resolve, reject) => {
+      const abortSignal = options && options.abortSignal;
+      let rejectOnAbort: Func<void, void>;
+      if (abortSignal) {
+        rejectOnAbort = () => {
+          const err = createAbortError();
+          log.error("[%s] [%s]", this.id, err.message);
+          return reject(err);
+        };
+        if (abortSignal.aborted) {
```
Comment (ramya-rao-a) [inline lib/connection.ts:544]: The early exit check is made towards the end just before deciding whether or not to add event listener to the abort signal. Now that we are doing the same here, we can remove that check

File: lib/connection.ts [hunk 3]
```diff
@@ -533,14 +533,49 @@ export class Connection extends Entity {
...
+        };
+
+        onAbort = () => {
+          removeListeners();
+          if (rheaSession.is_open()) {
+            // This scenario *shouldn't* be possible because if `is_open()` returns true,
+            // our `onOpen` handler should have executed and removed this abort listener.
+            // This is a 'just in case' check in case the operation was cancelled sometime
+            // between when the session's state was updated and when the sessionOpen
+            // event was emitted.
```
Comment (ramya-rao-a) [inline lib/connection.ts:552]: > and when the sessionOpen event was emitted

Nit: Isnt it more like when the sessionOpen event was handled?

**Review by ramya-rao-a — APPROVED** [review]

File: lib/connection.ts (hunk 3 above)
Comment (chradek) [reply lib/connection.ts:552]: That is true :smile:
